engine/test-src/org/pentaho/di/concurrency/SlaveServerConcurrentTest.java

```diff
@@ -0,0 +1,184 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2002-2016 by Pentaho : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.concurrency;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.pentaho.di.cluster.SlaveServer;
+import org.pentaho.di.repository.StringObjectId;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static junit.framework.Assert.assertEquals;
+
+/**
+ * We have a {@link SlaveServer} with a bunch of attributes (userName, password, port and others).
+ *
+ * The goal of this test is to verify that attributes from one SlaveServer doesn't mix with attributes of
+ * another SlaveServer in concurrent environment when performing replace and clone methods.
+ *
+ * We have a {@code sharedSlaveServer} that will be share across {@link Replacer} and {@link Getter}
+ *
+ * {@link Replacer} generates consistent slaveService's. By consistence the following is meant:
+ * Each slaveServer has it's id (random int) and all his fields are named in pattern: FIELD_NAME + slaveServerId.
+ * Then replacer replaces {@code sharedSlaveServer} with generated slaveServer.
+ *
+ * Note: it is expected that {@code sharedSlaveServer} will always be in consistent state (all his fields will end
+ * with the same id).
+ *
+ * And that's exactly what {@link Getter} does:
+ *
+ * It takes a snapshot of {@code sharedSlaveServer} and verifies it's consistency.
+ * We have to take a snapshot here because {@code sharedSlaveServer} can be updated in the middle of consistency
+ * verification. It's absolutely fine, cause it will be updated with consistent values.
+ *
+ */
+public class SlaveServerConcurrentTest {
```

Comment:
Have you checked your new r/w lock based class test on Oracla Java? Doesn't it suffer from the same memory leak as those ignored ones?